kernel: thread: fix lowest thread priority

jonas-5ae29c opened the pull request:
fixes #85153

## Alternative 1 - Using the lowest configured thread priority for the idle thread

```C
 #define K_HIGHEST_THREAD_PRIO (-CONFIG_NUM_COOP_PRIORITIES) 
 #define K_LOWEST_THREAD_PRIO (CONFIG_NUM_PREEMPT_PRIORITIES - 1)  // changed from CONFIG_NUM_PREEMPT_PRIORITIES 
 #define K_IDLE_PRIO K_LOWEST_THREAD_PRIO 
 #define K_HIGHEST_APPLICATION_THREAD_PRIO (K_HIGHEST_THREAD_PRIO) 
 #define K_LOWEST_APPLICATION_THREAD_PRIO (K_LOWEST_THREAD_PRIO - 1) 
```

This is consistent with the documentation found here [Threads - Thread Priorities](https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#thread-priorities), [System Threads](https://docs.zephyrproject.org/latest/kernel/services/threads/system_threads.html).

There are at most 128 cooperative and 128 preemptive thread priorities.
* cooperative threads: (-CONFIG_NUM_COOP_PRIORITIES) to -1
* preemptive threads: 0 to (CONFIG_NUM_PREEMPT_PRIORITIES - 1)
* idle thread has a preemptive thread priority: CONFIG_NUM_PREEMPT_PRIORITIES - 1

**TODO**
* [ ] Update documentation [CONFIG_NUM_PREEMPT_PRIORITIES](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_NUM_PREEMPT_PRIORITIES)
* [ ] Decide what priority the main thread should have [CONFIG_MAIN_THREAD_PRIORITY](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MAIN_THREAD_PRIORITY)

## Alternative 2 - Using a priority lower than the lowest configured thread priority for the idle thread

Restrict CONFIG_NUM_PREEMPT_PRIORITIES to the range 0 to 127 and let the idle thread priority be one priority lower than the lowest configured thread priority.

```C
 #define K_LOWEST_THREAD_PRIO CONFIG_NUM_PREEMPT_PRIORITIES // no change
 #define K_IDLE_PRIO K_LOWEST_THREAD_PRIO
```

There are at most 128 cooperative and 127 preemptive thread priorities and 1 idle thread priority.
* cooperative threads: (-CONFIG_NUM_COOP_PRIORITIES) to -1
* preemptive threads: 0 to (CONFIG_NUM_PREEMPT_PRIORITIES - 1)
* idle thread: CONFIG_NUM_PREEMPT_PRIORITIES

**TODO**
* [x] Update range [CONFIG_NUM_PREEMPT_PRIORITIES](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_NUM_PREEMPT_PRIORITIES)
* [x] Update documentation [Threads - Thread Priorities](https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#thread-priorities)

## Either way

* Update documentation; [System Threads](https://docs.zephyrproject.org/latest/kernel/services/threads/system_threads.html)
  * [ ] in docs: main thread has priority -1 when there are no preemptible threads. But is has priority -2 in that case https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MAIN_THREAD_PRIORITY
  * [x] in docs: The idle thread always uses the lowest configured thread priority and can be a cooperative thread.

cfriedt reviewed (dismissed):
This makes sense to me.

peter-mitsis commented:
Off by ones frequently induce some level of brain pain, but I am thinking that "Alternative 2" would be the better approach.

5ae29c commented:
I also just found the following

https://github.com/zephyrproject-rtos/zephyr/blob/a36c91b4cc49b9b54030121004a1565d6fc69475/kernel/idle.c#L77-L92

Currently (before any code of this PR) `CONFIG_PREEMPT_ENABLED` can be false because `CONFIG_NUM_PREEMPT_PRIORITIES` is 0, but the idle thread still has a preemptible priority of 0.

jonas-5ae29c committed:
kernel: thread: fix thread priority off-by-one error

Setting CONFIG_NUM_PREEMPT_PRIORITIES to 128 causes the idle thread to be
assigned priority 128, which exceeds the int8_t range. This results in the
idle thread being assigned the highest priority (-128) instead of the
lowest, causing threads to not wake up from k_sleep.

Restrict the range of CONFIG_NUM_PREEMPT_PRIORITIES to 0 to 127 to ensure
the idle thread always has the lowest priority.

Signed-off-by: Jonas Spinner <jonas.spinner@burkert.com>

jonas-5ae29c committed:
doc: kernel: threads: fix idle thread priority documentation

Clarify that the idle thread has priority CONFIG_NUM_PREEMPT_PRIORITIES,
not (CONFIG_NUM_PREEMPT_PRIORITIES - 1).

The idle thread priority ranges from 0 to 127 and never has a cooperative
thread priority.

Signed-off-by: Jonas Spinner <jonas.spinner@burkert.com>

5ae29c commented:
I'm quite happy with the current change set. It restricts the range of CONFIG_NUM_PREEMPT_PRIORITIES to 0 to 127 and changes some incorrect documentation regarding the idle thread priority.

This does not affect any actual priorities (user thread, main or idle thread) and only prevents using CONFIG_NUM_PREEMPT_PRIORITIES=128 which was broken anyway.

Remaining issues are discrepancies between [CONFIG_MAIN_THREAD_PRIORITY](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MAIN_THREAD_PRIORITY) and [System Threads](https://docs.zephyrproject.org/latest/kernel/services/threads/system_threads.html) about the value of the main thread priority. I left this out of this PR as it is a separate concern.

peter-mitsis reviewed (commented):
On the graphic ... is it possible to darken the text? Maybe it is just me, but the text is coming up to be very faint.

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -53,7 +53,7 @@ config NUM_PREEMPT_PRIORITIES
 	default 0 if !MULTITHREADING
 	default 15
-	range 0 128
+	range 0 127
```
But +128 can not be represented as a signed 8-bit value which goes from -128 to +127.

cfriedt replied:
Ah, I didn't realize that was a limitation. Makes sense now.

5ae29c commented:
> On the graphic ... is it possible to darken the text? Maybe it is just me, but the text is coming up to be very faint.

Good catch!
